### PR TITLE
New lint: default_numeric_fallback

### LINKED ISSUE #6064: Lint default numeric type fallback
### What it does

Lint any integer or floating point constants that have their type determined by the default fallback rather than inference

### Categories (optional)

- Kind: pedantic

*What is the advantage of the recommended code over the original code*

Avoids unintended use of the numeric default types, forcing the author to think about what type is appropriate.

### Drawbacks

Forcing the author to think about what type is appropriate.

### Example

```rust
let (x, y) = (5, 6.0);
```

Could be written as:

```rust
let (x, y): (i32, f32) = (5, 6.0);
let (x, y) = (5i32, 6.0f32);
```

### PR DESCRIPTION
**Pull request by Y-Nak**
fixes #6064 
r? @flip1995 

As we discussed in [here](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Issue.20.236064/near/224647188) and [here](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Issue.20clippy.236064/near/224746333),   I start implementing this lint from the strictest version.
In this PR, I'll allow the below two cases to pass the lint to reduce FPs.
 
1. Appearances of unsuffixed numeric literals in `Local` if `Local` has a type annotation, for example:
```rust
// Good.
let x: i32 = 1;

// Also good.
let x: (i32, i32) = if cond {
   (1, 2)
} else {
   (2, 3)
};
```

2. Appearances of unsuffixed numeric literals in args of `Call` or `MethodCall`  if corresponding arguments of their signature have concrete types, for example:
```rust
fn foo_mono(x: i32) -> i32 {
    x
}

fn foo_poly<T>(t: T) -> t {
    t
}

// Good.
let x = foo_mono(13);

// Still bad.
let x: i32 = foo_poly(13);
```

changelog: Added restriction lint: `default_numeric_fallback`

### CODE REVIEW

**Comment by Y-Nak** [comment]
@flip1995 I finished the first implementation, though CI fails (maybe) due to updating of the toolchain.
As we discussed, `hir_ty_to_ty` causes ICR if passed type includes `hir::TyKind::Infer` because it doesn't assume the use from inside of function bodies. I'll try to resolve this problem tomorrow.

**Comment by Y-Nak** [comment]
I added some tests mainly related to `hir::Tykind::Infer` and fix some bugs.
Now all tests have been passed! Could you review this PR, please?

**Comment by Y-Nak** [comment]
Changes: Rewrote `enclosing_body_owner_opt` to correctly determine if a target `hir_id` is included in a body and renamed it to `enclosing_body_opt`.

File: tests/ui/default_numeric_fallback.stderr [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+   |
+LL |     let x: (_, i32) = (1, 2);
+   |                        ^
+   |
+   = help: consider adding suffix to avoid default numeric fallback
+
+error: default numeric fallback might occur
+  --> $DIR/default_numeric_fallback.rs:41:27
+   |
+LL |     let x: (_, i32) = (1, 2);
```
Comment (flip1995) [inline tests/ui/default_numeric_fallback.stderr:61]: This is a FP, because the type of the second tuple item has been specified.

**Review by flip1995 — COMMENTED** [review]
Let me preface this with: I hate giving reviews like this, when so much work was put into this.

So I don't think we can walk _every_ type we find just to check if somewhere in the type some type has to be inferred. This is just way too overkill for this lint.

I also don't quite understand why this lint has to be so complicated. If I had to implement this, I would just implement it on a big match on the `ExprKind` inside the `check_expr` function. To avoid linting `let` statements where types are defined, you can just save the `HitId` of the `local.init` expression and then skip this expression in `check_expr`.

Looking at the tests, the only thing you want to lint are int or float literals. And the only case you don't want to lint is if you are inside a let binding where a type definition is in the let binding. You can achieve this with:

```rust
fn check_stmt(&mut self, cx: &LateContext<'_>, stmt: &Stmt<'_>) {
    if_chain! {
        if let StmtKind::Local(local) = stmt.kind;
        if local.ty.is_some();
        if let Some(init) = local.init;
        then {
            self.skip_lint.insert(init.hir_id);
        }
    }
}

fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
    let parent_iter = cx.tcx.hir().parent_iter(expr.hir_id);
    if_chain! {
        if let ExprKind::Lit(lit) = expr.kind;
        if matches!(lit.node, LitKind::Int(_, LitIntType::Unsuffixed) | LitKind::Float(_, LitFloatType::Unsuffixed));
        if let Some(ty) = cx.typeck_results().expr_ty(expr);
        if matches!(ty.kind, ty::Int(IntTy::I32) | ty::Float(FloatTy::F64));
        if !parent_iter.any(|(hir_id, _)| self.skip_lint.contains(hir_id));
        then {
            span_lint_and_sugg(..);
        }
    }
}
```

I didn't test this code, but I'm fairly confident that it should work like this or similar.

File: tests/ui/default_numeric_fallback.stderr (hunk 1 above)
Comment (Y-Nak) [reply tests/ui/default_numeric_fallback.stderr:61]: Let me explain.
Firstly, I thought we had agreed on how this lint should be used. First of all, I thought we both agreed on how the lint was to be used, in that it was to be used in situations where a very high level of reliability was required at the type level, and the appearance of FP was inevitable. Isn't that why we chose the category `restriction'?
The reason this lint is so long is not to detect the possibility of a `Default numeric fallback`, but to try to reduce FP as much as possible while not generating NP.
I suspect your implementation will produce more NP and possibly FP than my implementation.
For example, wouldn't it result in FP in the following case?
```rust
fn foo<T>(t:T) {}
let x: Foo = foo(13);
```

**Comment by Y-Nak** [comment]
Let me explain.
Firstly, I thought we had agreed on how this lint should be used. First of all, I thought we both agreed on how the lint was to be used, in that it was to be used in situations where a very high level of reliability was required at the type level, and the appearance of FP was inevitable. Isn't that why we chose the category `restriction'?
The reason this lint is so long is not to detect the possibility of a `Default numeric fallback`, but to try to reduce FP as much as possible while not generating NP.
I suspect your implementation will produce more NP and possibly FP than my implementation.
For example, wouldn't it result in FP in the following case?
````
fn foo<T>(t:T) {};
let x: Foo = foo(13);
```

**Comment by Y-Nak** [comment]
Also this case would cause FN in your implementation.
```rust 
fn foo() -> i32 {
    13
}
```

**Comment by Y-Nak** [comment]
Just in case: I don't care at all if I should remove all the lines I wrote, I'm sorry if I took up your time, as for me I learned a lot through this implementation.

**Comment by Y-Nak** [comment]
Anyway I'll rewrite this tomorrow. Thanks for reviewing.

**Comment by Y-Nak** [comment]
Changes: Simplify `DefaultNumericFallback` as @flip1995 suggested.

In fact, I can't yet be convinced that this implementation is correct.
Consider the next scenario. 

A user wrote a code like below.
```rust
fn foo(x: i32) -> i32 {}
fn main() {
    // FP!
    let x = foo(1);
}
```

The current implementation would lint this although that's FP, so the user added a type annotation to avoid FP.
```rust
fn foo(x: i32) -> i32 {}
fn main() {
    let x: _ = foo(1);
}
```

After months or so, the user noticed that they have to rewrite it so that  `foo` should also receive `u32` or other integer types because it turns out in some part of their code other integer types should be used with `foo` for some reasons.
So they rewrote it like below.
```rust
fn foo<T: Integer>(x: T) -> i32 {}
fn main() {
    // FN! Default numeric fallback has occured.
    let x: _ = foo(1);
}
```

This seems a disaster for me.
I can't understand why this kind of FPs/FNs should be acceptable.

The "complex" version can avoid this kind of disaster, and we can reduce the complexity in exchange for some increase of FPs or by allowing some FNs.
But I fully respect your works and experiences in this region, so I'm willing to accept changes if you still think this version is preferable to the first version.

**Comment by Y-Nak** [comment]
As a reference, I simplified the "complex" version and reduced FPs by allowing FNs.
[default_numeric_fallback.rs](https://github.com/Y-Nak/rust-clippy/blob/simplified-default-numeric-fallback/clippy_lints/src/default_numeric_fallback.rs)
[tests/ui/default_numeric_fallback.rs](https://github.com/Y-Nak/rust-clippy/blob/simplified-default-numeric-fallback/tests/ui/default_numeric_fallback.rs)
[tests/ui/default_numeric_fallback.stderr](https://github.com/Y-Nak/rust-clippy/blob/simplified-default-numeric-fallback/tests/ui/default_numeric_fallback.stderr)

**Comment by flip1995** [comment]
> Isn't that why we chose the category `restriction`?

No, the ``restriction`` group is the correct group for this lint, because fallback to the default int/float type is a Rust concept that does not really cause problems. Having to specify the type anyway would _restrict_  the language, so that you can't just write `x = 42` anymore without specifying the type. If we want no FNs without caring about FPs, we could just lint every literal that is ``Unsuffixed``

> For example, wouldn't it result in FN in the following case?
> 
> ```rust
> fn foo<T>(t:T) -> Foo {}
> let x: Foo = foo(13);
> ```

Yes, I guess it would. But I think this is a really constructed example, because you wouldn't specify a type here. Even if you change it to 

```rust
fn foo<T>(t:T) -> Foo<T> {}
let x: Foo<_> = foo(13);
```

I don't think this would be a FN, because you specified ``T`` to be ``_`` and therefore said "It doesn't matter".

> Also this case would cause FP in your implementation.
> 
> ```rust
> fn foo() -> i32 {
>     13
> }
> ```

Why is that a FP? The type is not immediately visible for the `13`, but defined via the return value. If I would care about int/float types that much, I would also like to make this clear in return positions.

I now start to think that our interpretations of what the lint should do is different. You say, that only places where actual fallback occurs should be linted. I think all places where no type is specified immediately in the same statement should be linted.

> so the user added a type annotation to avoid FP.

Adding `_` as a type is a solution to this lint, that could lead to the disaster you described. But we don't and shouldn't suggest that as the fix. Rather we're suggesting to add `i32`/`f64` to the literal as the fix, which is resistent to the disaster you described.

---

In the end all your work you put into this helped to produce the ICE reproducer, no one else managed to come up with. So your work was definitely not in vain.

File: tests/ui/default_numeric_fallback.rs [hunk 2]
```diff
@@ -1,99 +1,18 @@
 #![warn(clippy::default_numeric_fallback)]
 #![allow(unused)]
-#![allow(clippy::never_loop)]
-#![allow(clippy::no_effect)]
-#![allow(clippy::unnecessary_operation)]
-
-fn ret_i31() -> i32 {
-    23
-}
```
Comment (flip1995) [inline tests/ui/default_numeric_fallback.rs]: I think you should keep all the tests, so if someone else wants to enhance/modify the lint there are already tests in place, that can be used for reference. Also then I will see, how my suggested changes changed the behavior of the lint.

**Comment by Y-Nak** [comment]
> Yes, I guess it would. But I think this is a really constructed example, because you wouldn't specify a type here. Even if you change it to

Consider this case.
```rust
fn foo<T, U> foo(t: T) -> U {
...
}
let x: MyType = foo(1);
```
In this case, T and U are irrelevant, so I still think this is FN.

And I think there is another problem in the current implementation. If an init expression is `ExprKind::Block` and the `Local` has type annotation, the current implementation doesn't lint anything in the block
Example:
```rust
fn main() {
    let x: MyType = {
        // FN. 
        let y = 13;

        ...
    };
}
```

**Comment by Y-Nak** [comment]
Changes: In order to make the comparison easier, I pushed another version that addresses the problems I pointed out. This implementation is almost the same as [this](https://github.com/rust-lang/rust-clippy/pull/6662#issuecomment-774670572).
I suppose that this implementation somewhat meets the requirements of both of us. 
I don't stick to the implementation, so please feel free to reject it.

**Review by flip1995 — COMMENTED** [review]
> And I think there is another problem in the current implementation. If an init expression is `ExprKind::Block` and the `Local` has type annotation, the current implementation doesn't lint anything in the block

Yeah your right, that is really too restrictive.

The problem I have with making this lint too complex is that complexity makes bugs more likely and also makes it harder to fix bugs. And IMO this is a kind of lint, that should not require this complexity, when choosing the right heuristics. 

---

The current implementation idea LGTM. But I don't think you need an extra visitor, since `LateLintPass` already is a visitor. You can just push the arg types in `check_expr` and then pop them in `check_expr_post`. The arguments get visited  after the `{Method}Call` anyway. That way you also get rid of the restriction, that it only works if the call is on a `Path`.

---

Please also add back all the tests you already had and the examples you gave in the comments of the PR with a comment each, why it should (not) lint.

**Comment by Y-Nak** [comment]
> The problem I have with making this lint too complex is that complexity makes bugs more likely and also makes it harder to fix bugs.

Yeah, I completely agree.

> But I don't think you need an extra visitor, since LateLintPass already is a visitor.

I agree on this in almost all cases.  But in this case, I think we can't use it as a visitor because there is no `check_stmt_post` or `check_local_post` in `LateLintPass`.

> Also you have to be careful with calling fn_sig on Closures. This will ICE

Thanks for pointing it out! I didn't realize it.

> Please also add back all the tests you already had and the examples you gave in the comments of the PR with a comment each, why it should (not) lint.

Ok, I'll add them later.

**Comment by flip1995** [comment]
>  But in this case, I think we can't use it as a visitor because there is no check_stmt_post or check_local_post in LateLintPass.

Ah right.  Please add something like "This lint can only be allowed on function levels and above to the `Known Problems` section then?

**Comment by Y-Nak** [comment]
Changes:
1. Add fn_sig_opt to get fn signature from HirId
2. Add more tests and edit `known problems`.

**Comment by Y-Nak** [comment]
Changes: Handle struct constructor case.
I suppose that `struct ctor` should be handled in the same way as `Call` or `MethodCall`, but not sure.

File: clippy_lints/src/default_numeric_fallback.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                if !ty_bound.is_integral();
+                then {
+                    span_lint_and_help(
+                        self.cx,
+                        DEFAULT_NUMERIC_FALLBACK,
+                        lit.span,
+                        "default numeric fallback might occur",
+                        None,
+                        "consider adding suffix to avoid default numeric fallback",
+                    );
```
Comment (flip1995) [inline clippy_lints/src/default_numeric_fallback.rs:88]: You could make this `span_lint_and_sugg` with `Applicability::MaybeIncorrect` and suggest `<literal>_i32`/`<literal>_f64`.

**Review by flip1995 — APPROVED** [review]
This looks really good to me now. 

I think linting in return places is good, because normally the return place is not as close to the fn sig as in the tests, so specifying the type more locally makes sense to me. Not regarding correctness, but readability.

---

One small change left to make.

**Comment by Y-Nak** [comment]
Changes: Use `span_lint_and_sugg` instead of `span_lint_and_help`.

**Comment by flip1995** [comment]
Perfect! I think we found a good balance for this lint. Thanks for all your work!

@bors r+


> Changes: Use `span_lint_and_sugg` instead of `span_lint_and_help`.

I really like those comments summarizing your latest changes, btw :+1:

**Comment by Y-Nak** [comment]
@flip1995 Thanks for your patience in the long discussion. I really appreciate your help!

**Pull request merged**